crates/wasmtime/src/profiling.rs

```diff
@@ -0,0 +1,106 @@
+    profile: Profile,
+    process: fxprof_processed_profile::ProcessHandle,
+    thread: fxprof_processed_profile::ThreadHandle,
+    start: Instant,
+}
+
+impl GuestProfiler {
+    /// Begin profiling a new guest. When this function is called, the current
+    /// wall-clock time is recorded as the start time for the guest.
+    ///
+    /// The `module_name` parameter is recorded in the profile to help identify
+    /// where the profile came from.
+    ///
+    /// The `interval` parameter should match the rate at which you intend
+    /// to call `sample`. However, this is used as a hint and not required to
+    /// exactly match the real sample rate.
+    pub fn new(module_name: &str, interval: Duration) -> Self {
+        let mut profile = Profile::new(
+            module_name,
+            std::time::SystemTime::now().into(),
+            interval.into(),
+        );
+        let process = profile.add_process(module_name, 0, Timestamp::from_nanos_since_reference(0));
+        let thread = profile.add_thread(process, 0, Timestamp::from_nanos_since_reference(0), true);
+        let start = Instant::now();
+        Self {
+            profile,
+            process,
+            thread,
+            start,
+        }
+    }
+
+    /// Add a sample to the profile. This function collects a backtrace from
+    /// any stack frames associated with the given `store` on the current
+    /// stack. It should typically be called from a callback registered using
+    /// [`Store::epoch_deadline_callback()`](crate::Store::epoch_deadline_callback).
+    pub fn sample(&mut self, store: impl AsContext) {
+        let now = Timestamp::from_nanos_since_reference(
+            self.start.elapsed().as_nanos().try_into().unwrap(),
+        );
+
+        let trace = WasmBacktrace::force_capture(store);
+        // Samply needs to see the oldest frame first, but we list the newest
+        // first, so iterate in reverse.
+        let frames = Vec::from_iter(trace.frames().iter().rev().map(|frame| {
+            let frame = if let Some(name) = frame.func_name() {
+                let idx = self.profile.intern_string(name);
+                Frame::Label(idx)
+            } else {
+                // `module_offset` should always be set because we require
+                // `generate_address_map` to be enabled.
+                let offset = frame.module_offset().unwrap();
+                Frame::InstructionPointer(offset.try_into().unwrap())
+            };
+            FrameInfo {
+                frame,
+                category_pair: CategoryHandle::OTHER.into(),
+                flags: FrameFlags::empty(),
+            }
+        }));
+
+        self.profile
+            .add_sample(self.thread, now, frames.into_iter(), CpuDelta::ZERO, 1);
+    }
+
+    /// When the guest finishes running, call this function to write the
+    /// profile to the given `output`.
+    pub fn finish(&mut self, output: impl std::io::Write) -> Result<()> {
```

Comment:
If you call `finish` again it'll just update the thread end-time and write out the whole profile again, including any new samples. So it's probably not useful to add more samples after calling `finish`, but it should be harmless.

That said, I only did it this way because there's still an extra reference to the `GuestProfiler` in the epoch callback after the instance finishes, and I didn't want to have to clone the profile. But if I can make Alex's suggestion work, of keeping the `GuestProfiler` in the store instead of in an `Arc`, then that won't be an issue and I could make the change you're suggesting too.